stores/metadata.go

```diff
@@ -2932,51 +2931,36 @@ func (s *SQLStore) deleteObject(tx *gorm.DB, bucket string, path string) (int64,
 // deletion goes from largest to smallest. That's because the batch size is
 // dynamically increased and the smaller objects get the faster we can delete
 // them meaning it makes sense to increase the batch size over time.
-func (s *SQLStore) deleteObjects(ctx context.Context, bucket string, path string) (numDeleted int64, _ error) {
+func (s *SQLStore) deleteObjects(ctx context.Context, bucket string, path string) (deleted bool, _ error) {
 	batchSizeIdx := 0
 	for {
+		start := time.Now()
 		var duration time.Duration
-		var rowsAffected int64
-		if err := s.retryTransaction(ctx, func(tx *gorm.DB) error {
-			start := time.Now()
-			res := tx.Exec(`
-			DELETE FROM objects
-			WHERE id IN (
-				SELECT id FROM (
-					SELECT id FROM objects
-					WHERE object_id LIKE ? AND SUBSTR(object_id, 1, ?) = ? AND ?
-					ORDER BY size DESC
-					LIMIT ?
-					) tmp
-				)`,
-				path+"%", utf8.RuneCountInString(path), path, sqlWhereBucket("objects", bucket),
-				objectDeleteBatchSizes[batchSizeIdx])
-			if err := res.Error; err != nil {
-				return res.Error
-			}
-			// prune slabs if we deleted an object
-			rowsAffected = res.RowsAffected
-			if rowsAffected > 0 {
-				s.triggerSlabPruning()
+		var prune bool
+		if err := s.bMain.Transaction(ctx, func(tx sql.DatabaseTx) error {
+			d, err := tx.DeleteObjects(ctx, bucket, path, objectDeleteBatchSizes[batchSizeIdx])
+			if err != nil {
+				return err
 			}
-			duration = time.Since(start)
+			deleted = deleted || d
+			prune = d
 			return nil
 		}); err != nil {
-			return 0, fmt.Errorf("failed to delete objects: %w", err)
+			return false, fmt.Errorf("failed to delete objects: %w", err)
 		}
-
-		// if nothing got deleted we are done
-		if rowsAffected == 0 {
-			break
+		if prune {
```

Comment:
Why don't we `triggerSlabPruning` after the batch delete, and why do we need the extra `prune` variable. Can't we use `deleted`?

Reply:
Hm I guess now that we batch the slab pruning we can move it to the end. Still need a var that tells us if a specific batch deleted anything though to know when to interrupt the loop.